tests/runtime/filter_aws.c

```diff
@@ -0,0 +1,262 @@
+    struct flb_filter_aws_init_options *ops;
+    flb_sds_t name;
+    struct flb_aws_client_mock_request_chain *request_chain;
+    char *output = NULL;
+
+    request_chain = FLB_AWS_CLIENT_MOCK(
+        response(
+            expect(URI, "/latest/meta-data/tags/instance"),
+            expect(METHOD, FLB_HTTP_GET),
+            set(STATUS, 404),
+        ),
+    );
+    flb_aws_client_mock_configure_generator(request_chain);
+
+    client_generator = flb_aws_client_get_mock_generator();
+    ops = flb_calloc(1, sizeof(struct flb_filter_aws_init_options));
+    name = flb_sds_create_size(100);
+    name = flb_sds_printf(&name, "flb_test_aws_ec2_tags_404");
+    ops->client_generator = client_generator;
+    ops->name = name;
+
+    ctx = flb_create();
+
+    in_ffd = flb_input(ctx, (char *) "lib", NULL);
+    TEST_CHECK(in_ffd >= 0);
+    flb_input_set(ctx, in_ffd, "tag", "test", NULL);
+
+
+    /* Prepare output callback context*/
+    cb_data.cb = callback_test;
+    cb_data.data = NULL;
+
+    /* Lib output */
+    out_ffd = flb_output(ctx, (char *) "lib", (void *)&cb_data);
+    TEST_CHECK(out_ffd >= 0);
+    flb_output_set(ctx, out_ffd,
+                   "match", "*",
+                   "format", "json",
+                   NULL);
+
+    filter_ffd = flb_filter(ctx, (char *) "aws", ops);
+    TEST_CHECK(filter_ffd >= 0);
+    ret = flb_filter_set(ctx, filter_ffd, "match", "*", NULL);
+    TEST_CHECK(ret == 0);
+    ret = flb_filter_set(ctx, filter_ffd, "ec2_instance_id", "false", NULL);
+    TEST_CHECK(ret == 0);
+    ret = flb_filter_set(ctx, filter_ffd, "az", "false", NULL);
+    TEST_CHECK(ret == 0);
+    ret = flb_filter_set(ctx, filter_ffd, "tags_enabled", "true", NULL);
+    TEST_CHECK(ret == 0);
+
+    ret = flb_start(ctx);
+    TEST_CHECK(ret == 0);
+
+    bytes = flb_lib_push(ctx, in_ffd, p, strlen(p));
+    if (!TEST_CHECK(bytes > 0)) {
+        TEST_MSG("zero bytes were pushed\n");
+    }
+
+    flb_time_msleep(1000);
+
+    output = get_output();
+    if (output) {
+        char *result = strstr(output, "\"Name\":\"my_ec2_instance\"");
+        if (!TEST_CHECK(result == NULL)) {
+            TEST_MSG("output:%s\n", output);
+        }
+        result = strstr(output, "\"CUSTOMER_ID\":\"70ec5c04-3a6e-11ed-a261-0242ac120002\"");
+        if (!TEST_CHECK(result == NULL)) {
+            TEST_MSG("output:%s\n", output);
+        }
+        result = strstr(output, "hello, from my ec2 instance");
+        if (!TEST_CHECK(result != NULL)) {
+            TEST_MSG("output:%s\n", output);
+        }
+    }
+    else {
+        TEST_CHECK(false);
+        TEST_MSG("output is empty");
+    }
+
+    flb_stop(ctx);
+    flb_destroy(ctx);
+    flb_sds_destroy(name);
+    flb_free(ops);
+
+    set_output(NULL);
+    flb_aws_client_mock_destroy_generator();
+}
+
+TEST_LIST = {
+    {"aws_ec2_tags_present", flb_test_aws_ec2_tags_present},
+    {"aws_ec2_tags_404", flb_test_aws_ec2_tags_404},
```

Comment:
I know these cases might never happen in real world, but part of the point of these tests is just to cover all code paths and make sure that there are no memory leaks in weird failure scenarios. 

So how about test cases for:
- IMDS responds with error response but not 404
- IMDS gives tag list but then when plugin tries to request one of the tags it always gets 404 or error

These are not mandatory. Also @matthewfala if you have time would appreciate checking out this PR and tests.

Reply:
https://github.com/fluent/fluent-bit/pull/6065/commits/cd1e0e8a3bffb4dd65ed7771c6e67522397cda6d